contracts/RelayHub.sol

```diff
@@ -186,43 +181,54 @@ contract RelayHub is IRelayHub {
         require(relayRequest.relayData.gasPrice <= tx.gasprice, "Invalid gas price");
         require(externalGasLimit <= block.gaslimit, "Impossible gas limit");
 
-        // We now verify that the paymaster will agree to be charged for the transaction.
-        (vars.success, vars.recipientContext, vars.gasLimits) =
-            canRelay(relayRequest,
-                    externalGasLimit, signature, approvalData);
-
-        if (!vars.success) {
-            revertReason = GsnUtils.getError(vars.recipientContext);
-            emit TransactionRejectedByPaymaster(
-                workerToManager[msg.sender],
-                relayRequest.relayData.paymaster,
-                relayRequest.request.from,
-                relayRequest.request.to,
-                msg.sender,
-                vars.functionSelector,
-                revertReason);
-            return (vars.success, revertReason);
-        }
-
-        // From this point on, this transaction will not revert nor run out of gas, and the paymaster will be charged
-        // for the gas spent.
+        (vars.gasLimits, vars.maxPossibleGas) =
+             verifyGasLimits(relayRequest, externalGasLimit);
 
     {
+
         //How much gas to pass down to innerRelayCall. must be lower than the default 63/64
         // actually, min(gasleft*63/64, gasleft-GAS_RESERVE) might be enough.
-        uint innerGasLimit = gasleft()*63/64-gasReserve;
+        uint256 innerGasLimit = gasleft()*63/64-gasReserve;
+        vars.gasBeforeInner = gasleft();
+
+        uint256 _tmpInitialGas = innerGasLimit + externalGasLimit + gasOverhead + postOverhead;
 
         // Calls to the recipient are performed atomically inside an inner transaction which may revert in case of
         // errors in the recipient. In either case (revert or regular execution) the return data encodes the
         // RelayCallStatus value.
-        (, bytes memory relayCallStatus) = address(this).call{gas:innerGasLimit}(
-            abi.encodeWithSelector(RelayHub.innerRelayCall.selector, relayRequest, signature, vars.gasLimits,
-                innerGasLimit + externalGasLimit-gasleft() + gasOverhead + postOverhead, /*totalInitialGas*/
-                abi.decode(vars.recipientContext, (bytes)))
+        (bool success, bytes memory relayCallStatus) = address(this).call{gas:innerGasLimit}(
+            abi.encodeWithSelector(RelayHub.innerRelayCall.selector, relayRequest, signature, approvalData, vars.gasLimits,
+                _tmpInitialGas - gasleft(),
+                vars.maxPossibleGas
+                )
         );
+        vars.success = success;
+        vars.innerGasUsed = vars.gasBeforeInner-gasleft();
         (vars.status, vars.relayedCallReturnValue) = abi.decode(relayCallStatus, (RelayCallStatus, bytes));
     }
     {
+        if (!vars.success) {
+            //Failure cases where the PM doesn't pay
+            if ( (vars.innerGasUsed < vars.gasLimits.acceptanceBudget ) && (
```

Comment:
So is it 'acceptanceBudget' or 'paymasterPaysAbove'? The only thing worse then a bad name is two bad names for the same variable.